src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -956,11 +956,29 @@ where
                     let total_buffered: usize = untimestamped_batches.iter().map(|(_, b)| b.len()).sum();
                     let reclock_source_upper = timestamper.source_upper();
 
+                    // This is another subtle point. Normally we would be free to emit the
+                    // untimestamped messages in any order so long as the frontiers are respected.
+                    // Unfortunately, downstream decoding operators rely on the messages being
+                    // presented at the exact order produced even if the messages happen at the
+                    // same timestamp. This should be fixed, but until this happens we need to
+                    // preserve the order here.
+                    //
+                    // We will treat the vector of untimestamped batches as a flat list of messages
+                    // and we will compute the size of the reclockable prefix. The messages in this
+                    // prefix can be reclocked immediately and emitted in the original order.
```

Comment:
we can update this to mention its just an optimization, but its probably more important to merge-as-is!

Reply:
I forgot to amend the commit -.-. I had written a different comment